### PR TITLE
Feature/sc 114

### PR DESCRIPTION
**Pull request by DanCh11**

### CODE REVIEW

File: sem_covid/entrypoints/notebooks/topic_modeling/topic_modeling_wrangling/tfidf_vectorization.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import spacy
+import pandas as pd
+
+from sklearn.feature_extraction.text import TfidfVectorizer
+
+nlp = spacy.load("en_core_web_sm")
+stopwords = nlp.Defaults.stop_words
+
+
+def vectorize_documents(document: List) -> pd.DataFrame:
```
Comment (costezki) [inline sem_covid/entrypoints/notebooks/topic_modeling/topic_modeling_wrangling/tfidf_vectorization.py:13]: Don't we already have this defined elsewhere? 

Please standardise the operations and place them in a single place. Most importantly, reuse these operations.

File: tests/unit/test_data_cleaner.py [hunk 2]
```diff
@@ -74,3 +74,29 @@ def test_data_cleaner():
...
+def test_clean_text_from_specific_characters(transformed_pwdb_dataframe):
+    pwdb_descriptive_data = transformed_pwdb_dataframe['title'].map(str) + ' ' + \
+                            transformed_pwdb_dataframe['background_info_description'].map(str) + ' ' + \
+                            transformed_pwdb_dataframe['content_of_measure_description'].map(str) + ' ' + \
+                            transformed_pwdb_dataframe['use_of_measure_description'] + ' ' + \
+                            transformed_pwdb_dataframe['involvement_of_social_partners_description']
+    unused_characters = ["\\r", ">", "\n", "\\", "<", "''", "%", "...", "\'", '"', "(", "\n", "[", "]"]
+    clean_text = clean_text_from_specific_characters(pwdb_descriptive_data, unused_characters)
+
+    assert "\\r" not in clean_text
```
Comment (costezki) [inline tests/unit/test_data_cleaner.py:88]: this list of assertions shall be a single line assert

File: tests/unit/test_lda_modeling.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    assert gensim.corpora.dictionary.Dictionary == type(word_modeling.id2word)
+    assert list == type(word_modeling.corpus)
+    assert str == type(word_modeling.id2word[0])
+    assert "Hardship case fund" == word_modeling.id2word[0]
+    assert tuple == type(word_modeling.corpus[0][0])
+    assert (0, 1) == word_modeling.corpus[0][0]
+    assert gensim.models.ldamulticore.LdaMulticore == type(lda_training)
+    assert numpy.ndarray == type(lda_training.get_topics())
+    assert "Corona crisis" == lda_training.id2word.get(1)
+    assert 10 == lda_training.num_topics
```
Comment (costezki) [inline tests/unit/test_lda_modeling.py:37]: why 10? I don't see this parameter provided?

File: tests/unit/test_lda_modeling.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import numpy
+import spacy
+
+from sem_covid.services.sc_wrangling.data_cleaning import clean_text_from_specific_characters
+from sem_covid.entrypoints.notebooks.topic_modeling.topic_modeling_wrangling.lda_modeling import WordsModeling
+
+nlp = spacy.load("en_core_web_sm")
+
+
+def test_word_modeling(transformed_pwdb_dataframe):
```
Comment (costezki) [inline tests/unit/test_lda_modeling.py:12]: you perform the whole topic modellign here, not testing a specific function. Or you do? If sho what do you test here? 

what will happen when we change the corpus? or the language model?

File: tests/unit/test_lda_modeling.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    noun_phrases = [word.text for word in doc.noun_chunks]
+    separated_noun_phrases = [[nouns] for nouns in noun_phrases]
+
+    word_modeling = WordsModeling(separated_noun_phrases)
+    lda_training = word_modeling.lda_model_training()
+
+    assert gensim.corpora.dictionary.Dictionary == type(word_modeling.id2word)
+    assert list == type(word_modeling.corpus)
+    assert str == type(word_modeling.id2word[0])
+    assert "Hardship case fund" == word_modeling.id2word[0]
```
Comment (costezki) [inline tests/unit/test_lda_modeling.py:31]: how do you know?

File: tests/unit/test_lda_modeling.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    pwdb_descriptive_data = transformed_pwdb_dataframe['title'].map(str) + ' ' + \
+                            transformed_pwdb_dataframe['background_info_description'].map(str) + ' ' + \
+                            transformed_pwdb_dataframe['content_of_measure_description'].map(str) + ' ' + \
+                            transformed_pwdb_dataframe['use_of_measure_description'] + ' ' + \
+                            transformed_pwdb_dataframe['involvement_of_social_partners_description']
+    unused_characters = ["\\r", ">", "\n", "\\", "<", "''", "%", "...", "\'", '"', "(", "\n", "[", "]"]
+    clean_text = clean_text_from_specific_characters(pwdb_descriptive_data, unused_characters)
+    doc = nlp(clean_text)
+
+    noun_phrases = [word.text for word in doc.noun_chunks]
```
Comment (costezki) [inline tests/unit/test_lda_modeling.py:22]: this code can be in a fixture, no?

File: tests/unit/test_lda_modeling.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    assert "Hardship case fund" == word_modeling.id2word[0]
+    assert tuple == type(word_modeling.corpus[0][0])
+    assert (0, 1) == word_modeling.corpus[0][0]
+    assert gensim.models.ldamulticore.LdaMulticore == type(lda_training)
+    assert numpy.ndarray == type(lda_training.get_topics())
+    assert "Corona crisis" == lda_training.id2word.get(1)
+    assert 10 == lda_training.num_topics
+    assert 10 == lda_training.chunksize
+    assert 10 == lda_training.passes
+    assert 100 == lda_training.iterations
```
Comment (costezki) [inline tests/unit/test_lda_modeling.py:40]: how do you know? why?

File: tests/unit/test_topic_visualizer.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+class TestTopicVisualizer:
+
+    @pytest.fixture(autouse=True)
+    def _insert_transformed_pwdb_dataframe(self, transformed_pwdb_dataframe):
+        self._transformed_pwdb_dataframe = transformed_pwdb_dataframe
+
+    def _instantiate_topic_information(self):
+        pwdb_descriptive_data = self._transformed_pwdb_dataframe['title'].map(str) + ' ' + \
+                                self._transformed_pwdb_dataframe['background_info_description'].map(str) + ' ' + \
```
Comment (costezki) [inline tests/unit/test_topic_visualizer.py:22]: DO not transform in so many places the data. Do it once in a function then reuse that function.

File: tests/unit/test_topic_visualizer.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                                self._transformed_pwdb_dataframe['content_of_measure_description'].map(str) + ' ' + \
+                                self._transformed_pwdb_dataframe['use_of_measure_description'] + ' ' + \
+                                self._transformed_pwdb_dataframe['involvement_of_social_partners_description']
+
+        unused_characters = ["\\r", ">", "\n", "\\", "<", "''", "%", "...", "\'", '"', "(", "\n", "[", "]"]
+        clean_text = clean_text_from_specific_characters(pwdb_descriptive_data, unused_characters)
+        doc = nlp(clean_text)
+
+        noun_phrases = [word.text for word in doc.noun_chunks]
```
Comment (costezki) [inline tests/unit/test_topic_visualizer.py:31]: is that a word or a noun phrase? I'm confused.

File: tests/unit/test_topic_visualizer.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        separated_noun_phrases = [[nouns] for nouns in noun_phrases]
+
+        word_modeling = WordsModeling(separated_noun_phrases)
```
Comment (costezki) [inline tests/unit/test_topic_visualizer.py:34]: really words?

File: tests/unit/test_topic_visualizer.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+        lda_training = word_modeling.lda_model_training()
+
+        return TopicInformation(lda_training, separated_noun_phrases)
```
Comment (costezki) [inline tests/unit/test_topic_visualizer.py:37]: so that is an object of your own class?

File: tests/unit/test_topic_visualizer.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+import pandas
+
+from sem_covid.services.sc_wrangling.data_cleaning import clean_text_from_specific_characters
+from sem_covid.entrypoints.notebooks.topic_modeling.topic_modeling_wrangling.topic_visualizer import TopicInformation
+from sem_covid.entrypoints.notebooks.topic_modeling.topic_modeling_wrangling.lda_modeling import WordsModeling
+
+nlp = spacy.load("en_core_web_sm")
+
+
+class TestTopicVisualizer:
```
Comment (costezki) [inline tests/unit/test_topic_visualizer.py:14]: why a test class?

File: tests/unit/test_topic_visualizer.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+        noun_phrases = [word.text for word in doc.noun_chunks]
+        separated_noun_phrases = [[nouns] for nouns in noun_phrases]
+
+        word_modeling = WordsModeling(separated_noun_phrases)
+        lda_training = word_modeling.lda_model_training()
+
+        return TopicInformation(lda_training, separated_noun_phrases)
+
+    def test_topic_information_injection(self):
```
Comment (costezki) [inline tests/unit/test_topic_visualizer.py:39]: which testing framework do you use here? 
how the above fixtures work?

File: tests/unit/test_topic_visualizer.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+        assert gensim.corpora.dictionary.Dictionary == type(self._instantiate_topic_information().id2word)
+        assert list == type(self._instantiate_topic_information().corpus)
+        assert gensim.models.ldamulticore.LdaMulticore == type(self._instantiate_topic_information().lda_model)
+
+    def test_format_topic_sentence(self):
+        topic_visualizer = self._instantiate_topic_information().format_topic_sentences()
+
+        assert pandas.core.frame.DataFrame == type(topic_visualizer)
+        assert "Document Number" in topic_visualizer
```
Comment (costezki) [inline tests/unit/test_topic_visualizer.py:48]: these 5 asserts shall be a single one

File: tests/unit/test_topic_visualizer.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+        assert "Dominant Topic" in topic_visualizer
+        assert "Topic Percentage Contribution" in topic_visualizer
+        assert "Keywords" in topic_visualizer
+        assert "Text" in topic_visualizer
+        assert len(topic_visualizer) == 115
```
Comment (costezki) [inline tests/unit/test_topic_visualizer.py:53]: why 115? what is your basis for this?

File: tests/unit/test_topic_visualizer.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+        assert topic_visualizer['Text'][0] == ['Hardship case fund']
+        assert topic_visualizer['Text'][1] != "Corona crisis"
```
Comment (costezki) [inline tests/unit/test_topic_visualizer.py:55]: how do you know?

File: tests/unit/test_topic_visualizer.py [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+        top_topic = self._instantiate_topic_information().select_most_illustrative_sentence()
+
+        assert pandas.core.frame.DataFrame == type(top_topic)
+        assert "Document Number" in top_topic
+        assert "Topic Number" in top_topic
+        assert "Topic Percentage Contribution" in top_topic
+        assert "Keywords" in top_topic
+        assert "Representative Text" in top_topic
+        assert len(top_topic) == 10
+        assert set(top_topic['Topic Number']) == {0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0}
```
Comment (costezki) [inline tests/unit/test_topic_visualizer.py:70]: why?

**Review by costezki — APPROVED** [review]

File: sem_covid/entrypoints/notebooks/topic_modeling/topic_modeling_wrangling/lsa_transformer.py [hunk 18]
```diff
@@ -0,0 +1,20 @@
+
+from typing import List
+
+from sklearn.decomposition import TruncatedSVD
+
+
+def lsa_document_transformation(document_vectors: List, num_components: int) -> tuple:
```
[the remark this answers is not in the document]
Comment (DanCh11) [reply sem_covid/entrypoints/notebooks/topic_modeling/topic_modeling_wrangling/lsa_transformer.py:7]: 1. yes
2. sure thing

File: sem_covid/entrypoints/notebooks/topic_modeling/topic_modeling_wrangling/tfidf_vectorization.py (hunk 1 above)
Comment (DanCh11) [reply sem_covid/entrypoints/notebooks/topic_modeling/topic_modeling_wrangling/tfidf_vectorization.py:13]: yes, I found it. I'll replace the function

File: tests/unit/test_topic_visualizer.py (hunk 10 above)
Comment (DanCh11) [reply tests/unit/test_topic_visualizer.py:34]: because this class is used not only for noun phrases

File: tests/unit/test_lda_modeling.py (hunk 3 above)
Comment (DanCh11) [reply tests/unit/test_lda_modeling.py:37]: The parameter is provided inside the `WordModeling` class

**Pull request closed without merging**